base-R.qmd

```diff
@@ -0,0 +1,502 @@
+
+### Apply family
+
+Another way of saying "map a function over each element of a vector" is "apply a function over each element of a vector", and base R provides a family of functions known as the "apply" family because they all use apply as a suffix.
+Here we'll give you a quick overview of this family so you can recognize them in the wild.
+
+The most important member of this family is `lapply()`, which is very similar to `purrr::map()`[^base-r-3].
+In fact, because we don't use any of `map()`'s more advanced features, you can replace every `map()` call in @sec-iteration with `lapply()`.
+
+[^base-r-3]: It just lacks convenient features like progress bars and reporting which element caused the problem if there's an error.
+
+There's no exact base R equivalent to `across()` but you can get close by using `[` twice along with `lapply():`
+
+```{r}
+df <- tibble(a = 1, b = 2, c = "a", d = "b", e = 4)
+
+# First find numeric columns
+num_cols <- sapply(df, is.numeric)
+num_cols
+
+# Then transform each column with lapply() and replace the original values
+df[num_cols] <- lapply(df[num_cols], \(x) x * 2)
+df
+```
+
+The code above uses a new function, `sapply()`.
+It's similar to `lapply()` but it always tries to simplify the result, hence the `s` in its name, here producing a logical vector instead of a list.
+We don't recommend using it for programming, because the simplification can fail and give you an unexpected type, but it's usually fine for interactive use.
+
+Base R provides a stricter version of `sapply()` called `vapply()`, short for vector apply.
+It takes an additional argument that specifies the expected type which ensures that simplification occurs the same way regardless of the input.
+For example, we could replace the `sapply()` call above with this `vapply()` where we specific that we expect `is.numeric()` to return a logical vector of length 1:
+
+```{r}
+vapply(df, is.numeric, logical(1))
+```
+
+Another important member of the apply family is `tapply()` which computes a single grouped summary:
+
+```{r}
+diamonds |> 
+  group_by(cut) |> 
+  summarise(price = mean(price))
+
+tapply(diamonds$price, diamonds$cut, mean)
+```
+
+Unfortunately it returns its results in a named vector which requires some gymnastics if you want to collect multiple summaries into a data frame matched with the variable you're grouping by.
+(It's certainly possible to avoid this and just work with free floating vectors, but that generally just makes more work down the road).
+If you want to see how you might use `tapply()` or other base techniques to perform other grouped summaries, Hadley has collected a few techniques [in a gist](https://gist.github.com/hadley/c430501804349d382ce90754936ab8ec).
```

Comment:
My favorite `tapply()` feature is that when given factors to "group" by, you get one summary result per factor level, even if that factor level didn't appear in the actual data. Which made it useful in hardhat for `weighted_table()`
https://github.com/tidymodels/hardhat/blob/1e6774ce9b5db35ce36a6247cd9390cd23db75da/R/table.R#L126

Somewhat like `dplyr::group_by(.drop = FALSE)`